File: csharp/components/StrataCommon/ProducerConsumer/AbstractDispatcher.cs
//  ##########################################################################
//  # File Name: AbstractDispatcher.cs
//  # Copyright: 2016, Sapientia Systems, LLC.
//  ##########################################################################

using Strata.Common.Utility;
using System;

namespace Strata.Common.ProducerConsumer
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// 
    /// </summary>
    ///  
    public abstract
    class AbstractDispatcher<T>:
        IDispatcher<T>
    {
        private IMultiMap<ISelector<T>,IConsumer<T>> itsConsumers;

        protected
        AbstractDispatcher()
        {
            itsConsumers = new MultiMap<ISelector<T>,IConsumer<T>>();
        }

        public IDispatcher<T>
        AttachConsumer(IConsumer<T> consumer)
        {
            itsConsumers.Insert(consumer.GetSelector(),consumer);
            return this;
        }

        public IDispatcher<T>
        DetachConsumer(IConsumer<T> consumer)
        {
            itsConsumers.Remove(consumer.GetSelector(),consumer);
            return this;
        }

        public bool
        HasConsumer(IConsumer<T> consumer)
        {
            return itsConsumers.ContainsValue(consumer.GetSelector(),consumer);
        }

        public abstract void
        Route(T payload);

        public abstract void
        Broadcast(T payload);

        public abstract void
        StartDispatching();

        public abstract void
        StopDispatching();

        public abstract bool
        IsDispatching();

        protected IMultiMap<ISelector<T>,IConsumer<T>>
        GetConsumers()
        {
            return itsConsumers;
        }

    }
}

//  ##########################################################################

File: java/components/StrataFoundation/src/strata/foundation/producerconsumer/AbstractDispatcher.java
// ##########################################################################
// # File Name:	AbstractDispatcher.java
// #
// # Copyright:	2014, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataCommon Framework.
// #
// #   			The StrataCommon Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataCommon Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataCommon
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.foundation.producerconsumer;

import strata.foundation.utility.IMultiMap;
import strata.foundation.utility.MultiMap;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public abstract 
class AbstractDispatcher<T>
    implements IDispatcher<T>
{
    private IMultiMap<ISelector<T>,IConsumer<T>> itsConsumers;

    /************************************************************************
     * Creates a new {@code AbstractDispatcher}. 
     *
     */
    public 
    AbstractDispatcher()
    {
        itsConsumers = new MultiMap<ISelector<T>,IConsumer<T>>();
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public AbstractDispatcher<T> 
    attachConsumer(IConsumer<T> consumer)
    {
        itsConsumers.put( consumer.getSelector(),consumer );
        return this;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public AbstractDispatcher<T> 
    detachConsumer(IConsumer<T> consumer)
    {
        itsConsumers.remove( consumer.getSelector(),consumer );
        return this;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    hasConsumer(IConsumer<T> consumer)
    {
        return itsConsumers.containsValue(consumer.getSelector(),consumer);
    }

    /************************************************************************
     *  
     *
     * @return
     */
    protected IMultiMap<ISelector<T>,IConsumer<T>>
    getConsumers()
    {
        return itsConsumers;
    }

}

// ##########################################################################
